File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-12/07-spiralMatrix-ii.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];
        int startRow = 0;
        int endRow = n - 1;
        int startCol = 0;
        int endCol = n - 1;
        
        int count = 1;
        
        while (startRow <= endRow && startCol <= endCol) {
            for (int col = startCol; col < endCol + 1; col++) {
                matrix[startRow][col] = count;
                count++;
            }
            for (int row = startRow + 1; row < endRow + 1; row++) {
                matrix[row][endCol] = count;
                count++;
            }
            for (int col = endCol - 1; col > startCol - 1; col--) {
                matrix[endRow][col] = count;
                count++;
            }
            for (int row = endRow - 1; row > startRow; row--) {
                matrix[row][startCol] = count;
                count++;
            }
            startRow++; endRow--;
            startCol++; endCol--;
        }
        return matrix;
    }
}

File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-12/07-spiralMatrix-ii.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        matrix = [[None] * n for _ in range(n)]
        startRow, endRow = 0, n - 1
        startCol, endCol = 0, n - 1
        
        count  = 1
        
        while startRow <= endRow and startCol <= endCol:
            
            for col in range(startCol, endRow + 1):
                matrix[startRow][col] = count
                count += 1
            
            for row in range(startRow + 1, endRow + 1):
                matrix[row][endCol] = count
                count += 1
            
            for col in range(endCol - 1, startCol - 1, -1):
                matrix[endRow][col] = count
                count += 1
            
            for row in range(endRow - 1, startRow, -1):
                matrix[row][startCol] = count
                count += 1
            
            startRow, endRow = startRow + 1, endRow - 1
            startCol, endCol = startCol + 1, endCol - 1
        
        return matrix
